README.md

# react-svg-radar-chart


Generate SVG radar charts with React.

[![Build Status](https://travis-ci.org/Spyna/react-svg-radar-chart.svg?branch=master)](https://travis-ci.org/Spyna/react-svg-radar-chart)
[![Coverage Status](https://coveralls.io/repos/github/Spyna/react-svg-radar-chart/badge.svg?branch=master)](https://coveralls.io/github/Spyna/react-svg-radar-chart?branch=master)

![an example](./demo.gif)

Based on <a href="https://github.com/derhuerst/svg-radar-chart" target="_blank">svg-radar-chart</a>.

## Installing

```shell
npm install react-svg-radar-chart
```

## Demo and Examples

### Running demo

 could be found here: https://spyna.github.io/react-svg-radar-chart/

### Source code

 could be found here: https://github.com/Spyna/react-svg-radar-chart/tree/master/src/demo.

### Run locally



```shell
git clone https://github.com/Spyna/react-svg-radar-chart.git
cd react-svg-radar-chart && npm install;
npm start;

```

## Usage


Import the `RadarChart` component and the default css file.

```js
import RadarChart from 'react-svg-radar-chart';
import 'react-svg-radar-chart/build/css/index.css'
```

Create the dataset you want to draw

```js
const data = [
      {
        data: {
          battery: 0.7,
          design: .8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: 'blue' }
      },
      {
        data: {
          battery: 0.6,
          design: .85,
          useful: 0.5,
          speed: 0.6,
          weight: 0.7
        },
        meta: { color: 'red' }
      }
    ];

const captions = {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    };

```

Use the component:

```js
<RadarChart
    captions={captions}
    data={data}
    size={450}
  />

```

Where:

 * `captions` are the names that will be drawn on the svg.
 * `data` are the data that will be displayed.
 * `size` is the size of the svg in *pixels*

Putting all togheter

```js
import React from 'react';

import RadarChart from 'react-svg-radar-chart';
import 'react-svg-radar-chart/build/css/index.css'

class App extends React.Component {
  render() {
 	 const data = [
      {
        data: {
          battery: 0.7,
          design: .8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: 'blue' }
      },
      {
        data: {
          battery: 0.6,
          design: .85,
          useful: 0.5,
          speed: 0.6,
          weight: 0.7
        },
        meta: { color: 'red' }
      }
    ];

	const captions = {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    };

    return (
      <div>
        <RadarChart
            captions={{
              // columns
              battery: 'Battery Capacity',
              design: 'Design',
              useful: 'Usefulness',
              speed: 'Speed',
              weight: 'Weight'
            }}
            data={[
              // data
              {
                data: {
                  battery: 0.7,
                  design: .8,
                  useful: 0.9,
                  speed: 0.67,
                  weight: 0.8
                },
                meta: { color: '#58FCEC' }
              },
            ]}
            size={400}
          />
        </div>
    );
  }
}

export default App;

```

## Props


| Property | meaning | mandatory |
| --- | --- | --- |
| captions | *object* the label on the chart | **yes** |
| data | *array* the data to display | **yes** |
| options | *object* custom options for the chart, see below | no |
| size | *number* custom size, the default is 300 | no |


## Data format

the `data` property must be an array of object composed as following:

```js
const dataset1 = {
 data : {property1 : value1 , property2: value2, ..., propertyN: valueN},
 meta : { color: 'green'}
}
const data = [dataset1, dataset2, ..., datasetN]

```
The `dataset` object must have:
 * a property named `data`: an object with the data to display: `battery: 0.7`. The name of each property must be found in the `captions` object.
 * a property named `meta`: an object with a `color` or `class` property to set the color or the class in the chart.


## Options, styles and defaults.

You can:

* Customize the color of the chart.
* Use your own *stylesheet* to customize the style of the radar chart.
* Pass some *options* to the `RadarChart` component.


 ### Customize the color of the chart

 When generating the `data` object you can use a `meta` property to set the *color* and the *class* of the shape.

 ```js
const dataset1 = {
 data : {...},
 meta : { color: 'green'}
};
const dataset2 = {
 data : {...},
 meta : { class: 'my-custom-class'}
};
```
### Use your own *stylesheet* to customize the style of the radar chart.

In the demo the `'react-svg-radar-chart/build/css/index.css'` is imported, but you can define you own if needed.

If you want to use a custom stylesheet you have to implement these classes:

 * `.shape`: the shape on the radar chart. The deafult is: `fill-opacity: .3;`.

* `.shape:hover` : if you want some *hover* effects. The default is:  `fill-opacity: .65;z-index: 100;`.

* `.scale`: the *circles* in the radar chart. The default is: `fill: #FAFAFA; stroke: #999;stroke-width: .2`.

* `.axis`: the axis of the chart. The default is: `stroke: #555;	stroke-width: .2`.

* `.caption`: the text on the chart. The default is: `fill: #444;	font-weight: 400;	text-shadow: 1px 1px 0 #fff`.

### Pass some *options* to the `RadarChart` component

The `RadarChart` component accepts a property named **options**. The default options are:

```js
const noSmoothing = points => {
  let d = 'M' + points[0][0].toFixed(4) + ',' + points[0][1].toFixed(4);
  for (let i = 1; i < points.length; i++) {
    d += 'L' + points[i][0].toFixed(4) + ',' + points[i][1].toFixed(4);
  }
  return d + 'z';
};

const defaultOptions = {
  size: 200,
  axes: true, // show axes?
  scales: 3, // show scale circles?
  captions: true, // show captions?
  captionMargin: 10,
  dots: false, // show dots?
  zoomDistance: 1.2, // where on the axes are the captions?
  setViewBox: (options) => `-${options.captionMargin} 0 ${options.size + options.captionMargin * 2} ${options.size}`, // custom viewBox ?
  smoothing: noSmoothing, // shape smoothing function
  axisProps: () => ({ className: 'axis' }),
  scaleProps: () => ({ className: 'scale', fill: 'none' }),
  shapeProps: () => ({ className: 'shape' }),
  captionProps: () => ({
    className: 'caption',
    textAnchor: 'middle',
    fontSize: 10,
    fontFamily: 'sans-serif'
  }),
  dotProps: () => ({
    className: 'dot',
    mouseEnter: (dot) => { console.log(dot) },
    mouseLeave: (dot) => { console.log(dot) }
  }),
  rotation: 0
};
```

| Property | meaning |
| --- | --- |
| size | the size in pixels of the svg chart, default is 300, can use only if `size` props is not used |
| axes | *true*/*false* show axes |
| scales | *number* how many scale circles |
| captions | *true*/*false* show captions |
| captionMargin | *number* The margin of the svg, to fit captions |
| dots | *true*/*false* show dots |
| zoomDistance | the distance of the zoom: 0.2 = closest |
| smoothing | the smoothing function |
| setViewBox | a function that take *options*' object as argument and returns viewBox as a string. ViewBox will be added to the *svg*|
| axisProps | a function that take the *current caption name* as arguments and returns an object. All the property will be added to the *axis svg component*|
| scaleProps | a function that take the *value of the scales* as arguments and returns an object. All the property will be added to the *scale svg component*|
| shapeProps | a function that take the *meta of the data* as arguments and returns an object. All the property will be added to the *shape svg component*|
| captionProps |  a function that take the *current caption name* as arguments and returns an object. All the property will be added to the *text svg component*|
| dotProps |  a function that takes the *dot svg component* as arguments and returns an object. All properties will be added to the *dot svg component*. The `mouseEnter` function returns useful information like `key`, `value` and the index of the shape the dot is associated with. See the demo as example for using the information as a tooltip|
| rotation | default `0`: The rotation (clockwise) in degrees. A number that indicates the rotation of the shape and captions. It must be beetween 0 and 360. Eg: `180` makes the chart upside down. `90` rotates the chart of 90 degrees. |



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.3.0](https://github.com/Spyna/react-svg-radar-chart/compare/v1.2.0...v1.3.0) (2021-06-05)


### Features

* add chart rotation ([bbd3806](https://github.com/Spyna/react-svg-radar-chart/commit/bbd3806775391c77fc9475e71f2ad583aa420100))

# Changelog


src/lib/index.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import RadarChart from './index';

it('exports the module', () => {
  expect(RadarChart).not.toBeNull();
});


src/lib/components/RadarChart.test.js
import React from 'react';
import ReactDOM from 'react-dom';
import RadarChart from './RadarChart';

it('RadarChart renders without crashing', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      }
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  ReactDOM.render(<RadarChart captions={captions} data={data} />, div);
});

it('renders the chart with a specified size', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      }
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };
  const size = '800';
  ReactDOM.render(
    <RadarChart captions={captions} data={data} size={size} />,
    div
  );
  expect(div.querySelector('svg').getAttribute('width')).toBe(size);
});

it('RadarChart renders with custom options', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      }
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  const options = {
    captions: false,
    axes: false,
    scales: 0,
    captionProps: () => ({
      className: 'caption',
      textAnchor: 'middle',
      fontFamily: 'sans-serif'
    })
  };

  ReactDOM.render(
    <RadarChart captions={captions} data={data} options={options} />,
    div
  );
});

it('RadarChart renders without axes', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      }
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  const options = {
    captions: false,
    axes: false
  };

  ReactDOM.render(
    <RadarChart captions={captions} data={data} options={options} />,
    div
  );
  expect(div.querySelectorAll('.axis').length).toBe(0);
});

it('RadarChart renders without captions', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      },
      meta: {}
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  ReactDOM.render(<RadarChart data={data} captions={captions} />, div);
});

it('RadarChart crashes when data are not in the right format', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 'wrong value',
        speed: 0.67
      }
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };
  expect(() => {
    ReactDOM.render(<RadarChart captions={captions} data={data} />, div);
  }).toThrow();
});

it('RadarChart crashes when captions are in the wrong format', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      }
    }
  ];
  let captions = 'with a wrong value';
  expect(() => {
    ReactDOM.render(<RadarChart captions={captions} data={data} />, div);
  }).toThrow();
  captions = { value: 'wrong' };
  expect(() => {
    ReactDOM.render(<RadarChart captions={captions} data={data} />, div);
  }).toThrow();
});

it('RadarChart crashes when captions are in the wrong format', () => {
  const div = document.createElement('div');

  const data = 'invalid';
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };
  expect(() => {
    ReactDOM.render(<RadarChart captions={captions} data={data} />, div);
  }).toThrow();
});

it('RadarChart renders with 2 dots', () => {
  const div = document.createElement('div');

  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      },
      meta: {}
    }
  ];
  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  const options = {
    captions: false,
    axes: false,
    dots: true,
    scales: 0,
    captionProps: () => ({
      className: 'caption',
      textAnchor: 'middle',
      fontFamily: 'sans-serif'
    }),
    dotProps: () => ({
      className: 'testDot',
      mouseEnter: dot => dot
    })
  };

  ReactDOM.render(
    <RadarChart captions={captions} data={data} options={options} />,
    div
  );

  expect(div.querySelectorAll('.testDot').length).toBe(2);
});

it('render with custom viewBox', () => {
  const div = document.createElement('div');
  const viewBox = '0 0 500 500';
  const setViewBox = () => viewBox;
  const data = [
    {
      data: {
        battery: 0.7,
        speed: 0.67
      },
      meta: {}
    }
  ];

  const captions = {
    battery: 'Battery Capacity',
    speed: 'Speed'
  };

  const options = {
    setViewBox
  };

  ReactDOM.render(
    <RadarChart captions={captions} data={data} options={options} />,
    div
  );

  expect(div.querySelectorAll(`svg[viewBox='${viewBox}']`).length).toBe(1);
});


src/demo/App.js
import React, { Component } from 'react';
import RadarChart from '../lib';

import './App.css';
import content from './demo-data.js';

function downloadURI(uri, name) {
  let link = document.createElement('a');
  link.download = name;
  link.href = uri;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
}

class App extends Component {
  constructor(props) {
    super(props);
    this.mousePos = [0, 0];
    this.state = {
      content,
      dot: {}
    };
  }

  componentDidMount() {
    document.addEventListener('mousemove', this.handleMouseMove);
  }

  // This is only used for the "with dots" example for
  // positioning the tooltip
  handleMouseMove = e => {
    this.mousePos = [e.pageX, e.pageY];
  };

  handleToolTip = dot => {
    this.setState({ dot });
  };

  downloadImage = imageId => event => {
    const imageElement = document.getElementById(imageId).outerHTML;
    const encoded = encodeURIComponent(imageElement);
    const header = 'data:image/svg+xml,';
    const dataUrl = header + encoded;
    downloadURI(dataUrl, `image-${imageId}.svg`);
  };

  render() {
    const { content, dot } = this.state;
    return (
      <div className="container">
        <h2>
          <a
            href="https://github.com/Spyna/react-svg-radar-chart"
            title="Github repo of this project"
          >
            <img
              src="./icon.png"
              width="100"
              alt="svg radar chart created with React Svg Radar Chart "
            />
          </a>
        </h2>
        <h1>React Svg Radar Chart</h1>
        <p>Generate SVG radar charts with React.</p>
        <p>
          <strong>react-svg-radar-chart</strong> is a lightweight library to
          create radar charts with React.
        </p>
        <div
          style={{
            display: 'flex',
            flexWrap: 'wrap',
            flexDirection: 'inherit',
            justifyContent: 'center'
          }}
        >
          {content.map((data, index) => (
            <div className="chart-container" key={`${index}-${data.name}`}>
              <h2>{data.name}</h2>
              <button
                className="download-button"
                onClick={this.downloadImage(`chart-${index}`)}
              >
                download image
              </button>
              <br />
              <RadarChart
                captions={data.captions}
                data={data.chart}
                options={data.options}
                size={data.size}
                id={`chart-${index}`}
              />
              {dot.key && (
                <div
                  className="tooltip"
                  style={{ left: this.mousePos[0], top: this.mousePos[1] }}
                >
                  <span>{`${dot.key}: ${dot.value}`}</span>
                  <br />
                  <span>{`Num Chart: ${dot.idx}`}</span>
                </div>
              )}
              <div className="source-container">
                <h3>data</h3>
                <pre>data: {JSON.stringify(data.chart, null, ' ')}</pre>
              </div>
            </div>
          ))}
        </div>
        <a href="https://github.com/Spyna/react-svg-radar-chart">
          <img
            style={{ position: 'absolute', top: 0, left: 0, border: 0 }}
            src="https://s3.amazonaws.com/github/ribbons/forkme_left_green_007200.png"
            alt="Fork me on GitHub"
          />
        </a>
      </div>
    );
  }
}

export default App;


src/lib/components/radar.js
import React from 'react';

const polarToX = (angle, distance) => Math.cos(angle - Math.PI / 2) * distance;

const polarToY = (angle, distance) => Math.sin(angle - Math.PI / 2) * distance;

const points = points => {
  return points
    .map(point => point[0].toFixed(4) + ',' + point[1].toFixed(4))
    .join(' ');
};

const axis = options => (col, i) => (
  <polyline
    key={`poly-axis-${i}`}
    points={points([
      [0, 0],
      [
        polarToX(col.angle, options.chartSize / 2),
        polarToY(col.angle, options.chartSize / 2)
      ]
    ])}
    {...options.axisProps(col)}
  />
);

const dot = (columns, options) => (chartData, i) => {
  const data = chartData.data;
  const meta = chartData.meta || {};
  const extraProps = options.dotProps(meta);
  let extraPropsSvg = {};
  let mouseEnter = () => {};
  let mouseLeave = () => {};
  if (extraProps.mouseEnter) {
    mouseEnter = extraProps.mouseEnter;
  }
  if (extraProps.mouseLeave) {
    mouseLeave = extraProps.mouseLeave;
  }
  if (extraProps.r) {
    extraPropsSvg.r = extraProps.r;
  }
  if (extraProps.fill) {
    extraPropsSvg.fill = extraProps.fill;
  }
  if (extraProps.stoke) {
    extraPropsSvg.stroke = extraProps.stroke;
  }
  if (extraProps.stokeWidth) {
    extraPropsSvg.strokeWidth = extraProps.strokeWidth;
  }
  return columns.map(col => {
    const val = data[col.key];
    if ('number' !== typeof val) {
      throw new Error(`Data set ${i} is invalid.`);
    }

    return (
      <circle
        key={`dot-${col.key}-${val}`}
        cx={polarToX(col.angle, (val * options.chartSize) / 2)}
        cy={polarToY(col.angle, (val * options.chartSize) / 2)}
        className={[extraProps.className, meta.class].join(' ')}
        onMouseEnter={() => mouseEnter({ key: col.key, value: val, idx: i })}
        onMouseLeave={() => mouseLeave({})}
        {...extraPropsSvg}
      />
    );
  });
};

const shape = (columns, options) => (chartData, i) => {
  const data = chartData.data;
  const meta = chartData.meta || {};
  const extraProps = options.shapeProps(meta);
  let extraPropsSvg = {};
  if (!meta.fill) {
    meta.fill = meta.color;
  }
  if (meta.strokeWidth) {
    extraPropsSvg.strokeWidth = meta.strokeWidth;
  }
  if (meta.strokeDasharray) {
    extraPropsSvg.strokeDasharray = meta.strokeDasharray;
  }
  if (meta.strokeLinecap) {
    extraPropsSvg.strokeLinecap = meta.strokeLinecap;
  }
  return (
    <path
      key={`shape-${i}`}
      d={options.smoothing(
        columns.map(col => {
          const val = data[col.key];
          if ('number' !== typeof val) {
            throw new Error(`Data set ${i} is invalid.`);
          }

          return [
            polarToX(col.angle, (val * options.chartSize) / 2),
            polarToY(col.angle, (val * options.chartSize) / 2)
          ];
        })
      )}
      {...extraProps}
      {...extraPropsSvg}
      stroke={meta.color}
      fill={meta.fill}
      className={[extraProps.className, meta.class].join(' ')}
    />
  );
};

const scale = (options, value) => (
  <circle
    key={`circle-${value}`}
    cx={0}
    cy={0}
    r={(value * options.chartSize) / 2}
    {...options.scaleProps(value)}
  />
);

const caption = options => col => (
  <text
    key={`caption-of-${col.key}`}
    x={polarToX(col.angle, (options.size / 2) * 0.95).toFixed(4)}
    y={polarToY(col.angle, (options.size / 2) * 0.95).toFixed(4)}
    dy={(options.captionProps(col).fontSize || 10) / 2}
    {...options.captionProps(col)}
  >
    {col.caption}
  </text>
);

const render = (captions, chartData, options = {}) => {
  if ('object' !== typeof captions || Array.isArray(captions)) {
    throw new Error('caption must be an object');
  }
  if (!Array.isArray(chartData)) {
    throw new Error('data must be an array');
  }
  options.chartSize = options.size / options.zoomDistance;

  const { rotation = 0 } = options;

  const columns = Object.keys(captions).map((key, i, all) => {
    const angle = (Math.PI * 2 * i) / all.length + rotation * (Math.PI / 180);
    return {
      key,
      caption: captions[key],
      angle: angle
    };
  });
  const groups = [
    <g key={`g-groups}`}>{chartData.map(shape(columns, options))}</g>
  ];
  if (options.captions) {
    groups.push(<g key={`poly-captions`}>{columns.map(caption(options))}</g>);
  }
  if (options.dots) {
    groups.push(<g key={`g-dots`}>{chartData.map(dot(columns, options))}</g>);
  }
  if (options.axes) {
    groups.unshift(<g key={`group-axes`}>{columns.map(axis(options))}</g>);
  }
  if (options.scales > 0) {
    const scales = [];
    for (let i = options.scales; i > 0; i--) {
      scales.push(scale(options, i / options.scales));
    }
    groups.unshift(<g key={`poly-scales`}>{scales}</g>);
  }
  const delta = (options.size / 2).toFixed(4);
  return <g transform={`translate(${delta},${delta})`}>{groups}</g>;
};

export default render;


src/lib/components/RadarChart.js
import React from 'react';
import '../radar.css';
import radar from './radar';

const noSmoothing = points => {
  let d = 'M' + points[0][0].toFixed(4) + ',' + points[0][1].toFixed(4);
  for (let i = 1; i < points.length; i++) {
    d += 'L' + points[i][0].toFixed(4) + ',' + points[i][1].toFixed(4);
  }
  return d + 'z';
};

const setViewBox = options =>
  `-${options.captionMargin} 0 ${options.size + options.captionMargin * 2} ${
    options.size
  }`;

const defaultOptions = {
  size: 300,
  axes: true, // show axes?
  scales: 3, // show scale circles?
  captions: true, // show captions?
  dots: false, // show dots?
  zoomDistance: 1.2, // where on the axes are the captions?
  smoothing: noSmoothing, // shape smoothing function
  captionMargin: 10,
  setViewBox,
  axisProps: () => ({ className: 'axis' }),
  scaleProps: () => ({ className: 'scale', fill: 'none' }),
  shapeProps: () => ({ className: 'shape' }),
  dotProps: () => ({
    className: 'dot'
  }),
  captionProps: () => ({
    className: 'caption',
    textAnchor: 'middle',
    fontSize: 10,
    fontFamily: 'sans-serif'
  }),
  rotation: 0
};

const RadarChart = props => {
  const { data, captions, options, size = defaultOptions.size, id } = props;

  const chartOptions = {
    ...defaultOptions,
    ...options,
    size
  };

  const { setViewBox } = chartOptions;
  const chart = radar(captions, data, chartOptions);

  return (
    <svg
      version="1"
      xmlns="http://www.w3.org/2000/svg"
      width={size}
      height={size}
      viewBox={setViewBox(chartOptions)}
      id={id}
    >
      {chart}
    </svg>
  );
};

export default RadarChart;


src/demo/demo-data.js
const demoData = [
  {
    name: 'with color',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness'
    },
    chart: [
      // data
      {
        data: { battery: 0.7, design: 1, useful: 0.9 },
        meta: { color: '#edc951' }
      },
      {
        data: { battery: 1, design: 0.6, useful: 0.8 },
        meta: { color: '#cc333f' }
      },
      {
        data: { battery: 0.8, design: 0.7, useful: 0.6 },
        meta: { color: '#00a0b0' }
      }
    ],
    options: {
      scaleProps: () => ({
        fill: '#fafafa',
        stroke: '#999',
        strokeWidth: '.2'
      }),
      rotation: -1
    }
  },
  {
    name: 'with class name',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness'
    },
    chart: [
      // data
      {
        data: { battery: 0.7, design: 1, useful: 0.9 },
        meta: { class: 'iphone' }
      },
      {
        data: { battery: 1, design: 0.6, useful: 0.8 },
        meta: { color: '#cc333f' }
      },
      {
        data: { battery: 0.8, design: 0.7, useful: 0.6 },
        meta: { color: '#00a0b0' }
      }
    ]
  },
  {
    name: 'with more data',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 1,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      },
      {
        data: {
          battery: 1,
          design: 0.6,
          useful: 0.8,
          speed: 0.9,
          weight: 0.7
        },
        meta: { color: '#cc333f' }
      },
      {
        data: {
          battery: 0.8,
          design: 0.7,
          useful: 0.6,
          speed: 0.8,
          weight: 0.6
        },
        meta: { color: '#00a0b0' }
      }
    ]
  },
  {
    name: 'with dots',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 1,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      },
      {
        data: {
          battery: 1,
          design: 0.6,
          useful: 0.8,
          speed: 0.9,
          weight: 0.7
        },
        meta: { color: '#cc333f' }
      },
      {
        data: {
          battery: 0.8,
          design: 0.7,
          useful: 0.6,
          speed: 0.8,
          weight: 0.6
        },
        meta: { color: '#00a0b0' }
      }
    ],
    options: {
      dots: true,
      dotProps: () => ({
        className: 'dot',
        mouseEnter: this.handleToolTip,
        mouseLeave: this.handleToolTip
      })
    }
  },
  {
    name: 'with single data',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 0.8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      }
    ]
  },
  {
    name: 'with custom options',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 0.8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      }
    ],
    options: {
      axes: true, // show axes?
      scales: 5, // how many circles?
      captions: false, // show captions?
      zoomDistance: 0.75, // 0.1 to 1?
      axisProps: () => ({ className: 'custom-axis' }),
      scaleProps: () => ({ className: 'custom-scale', fill: 'none' }),
      shapeProps: () => ({ className: 'custom-shape' })
    }
  },
  {
    name: 'more with custom options',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 0.8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      }
    ],
    options: {
      axes: false, // show axes?
      scales: 8, // how many circles?
      captions: true, // show captions?
      captionMargin: 12
    }
  },
  {
    name: 'with custom size',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness',
      speed: 'Speed',
      weight: 'Weight'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.7,
          design: 0.8,
          useful: 0.9,
          speed: 0.67,
          weight: 0.8
        },
        meta: { color: '#edc951' }
      }
    ],
    size: 450
  },
  {
    name: 'With 180 degrees rotation',
    captions: {
      // columns
      battery: 'Battery Capacity',
      design: 'Design',
      useful: 'Usefulness'
    },
    chart: [
      // data
      {
        data: {
          battery: 0.8,
          design: 0.7,
          useful: 0.6
        },
        meta: { color: '#edc951' }
      }
    ],
    options: {
      scaleProps: () => ({
        fill: '#fafafa',
        stroke: '#999',
        strokeWidth: '.6'
      }),
      rotation: 180
    }
  }
];

export default demoData;
